balancer_wrapper.go

```diff
@@ -256,17 +256,20 @@ type acBalancerWrapper struct {
 	ccb           *ccBalancerWrapper // read-only
 	stateListener func(balancer.SubConnState)
 
-	mu        sync.Mutex
-	producers map[balancer.ProducerBuilder]*refCountedProducer
+	producersMu sync.Mutex
+	producers   map[balancer.ProducerBuilder]*refCountedProducer
 }
 
 // updateState is invoked by grpc to push a subConn state update to the
 // underlying balancer.
-func (acbw *acBalancerWrapper) updateState(s connectivity.State, curAddr resolver.Address, err error, readyChan chan struct{}) {
+func (acbw *acBalancerWrapper) updateState(s connectivity.State, curAddr resolver.Address, err error) {
 	acbw.ccb.serializer.TrySchedule(func(ctx context.Context) {
 		if ctx.Err() != nil || acbw.ccb.balancer == nil {
 			return
 		}
+		// Invalidate all producers on any state change.
+		acbw.closeProducers()
```

Comment:
What is the rationale behind closing producers for every state change? This will be a no-op until the subchannel reaches `Ready`, and actually is useful only when the subchannel transitions out of `Ready`, right? For example: what would closing producers do when the subchannel moves from TF to Idle?

Reply:
The balancer_wrapper (acbw) ensures we only call `close` on the producer once.

Reply:
> What is the rationale behind closing producers for every state change?

Just simplicity.  We could make it do it only when leaving READY, but then in theory an LB policy could create one in CONNECTING and it _could_ work depending on races (i.e. the subchannel becomes READY before the stream is created).  This way we know it can't deliver messages to one producer before entering ready or after leaving ready.  Willing to change this if you have a better idea.